py2many/clike.py

```diff
@@ -225,13 +225,13 @@ def _typename_from_annotation(self, node, attr="annotation") -> str:
                 node.container_type = type_node.container_type
                 return self._visit_container_type(type_node.container_type)
             if typename is None:
-                raise Exception(f"Could not infer: {type_node}")
+                raise NotImplementedError(f"Could not infer: {type_node}")
```

Comment:
This one could be `TypeError` with `node.lineno`. I'm sure there is a format we can report the errors in that allows IDEs and other tools to implement "click to go to the error". But that can happen in the code that catches the exception.

Reply:
It is the transpiler logic is not working properly, iirc this was Julia only which was triggering this for lots of pyjl/ py2many/ inputs.

Implementing a custom subclass of `NotImplementedError` which has extra attributes for context would be useful, but is beyond the planned scope of this PR.  When building that, the implementation should also consume that extra context information in a meaningful way in order to ensure it is designed correctly.

Reply:
Julia Call in pyjl/transpiler.py and Go Return in pygo/transpiler.py trigger this.  In both cases `type_node` is "None".

I can see why you think `TypeError` is appropriate, but the logic in those transpilers is not "wrong" - it isn't using the wrong type when calling this function.  The calling logic is just not able to handle more complex scenarios.

The problem with raising normal python exceptions for any of these types of problems is then we need to mask them, and real uses of those exceptions, caused by bad python code, will also be masked, allowing regressions.

In the end, all of these `NotImplementedError` need to be eradicated.  In some cases, like here, they should be replaced with `assert ..` or even `TypeError`.

In addition we should create a special subclass of `NotImplementedError` for each scenario which are not "ever" going to be implemented, such as
- syntax which is not possible in the target language, 
- syntax we dont intend to support for any language, etc